daisy/step_create_instances.go

```diff
@@ -51,73 +57,46 @@ func (ci *CreateInstances) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
-func logSerialOutput(ctx context.Context, s *Step, ii InstanceInterface, ib *InstanceBase, port int64, interval time.Duration) {
+func logSerialOutput(s *Step, name string, watcher *SerialOutputWatcher, wcProvider func() io.WriteCloser) {
 	w := s.w
 	w.stepWait.Add(1)
+	// The buffered channel ensures that we don't lose serial output if we have delays uploading
+	// the previous chunk to GCS. Assuming a poll interval of 3 seconds, the buffer of size 1024
+	// allows ~51 minutes of backlog before we start losing logs.
+	c := make(chan string, 1024)
+	(*watcher).Watch(name, serialPortToArchive, c, serialPortPollInterval)
+	(*watcher).start(name)
```

Comment:
I'm glad you brought this up, as the current approach feels fragile to me as well. Any ideas on where that central place might be?

Reply:
Since serial port output can start as soon as an instance is created, it makes sense to have `start()` in step_create_instance.go. I'm not sure we can have `start()` somewhere else and ensure no serial output is lost.

Because of this, in the future, we might have subscribers call `Watch()` after `start()`. If we do allow this (currently it panics), we can handle it in a couple of ways:
* buffer output on the watcher side (up to a point to allow some later subcribers, but not have to store whole output for the duration of the execution), or
* allow subcribers to know if output has already started for given port/instance name before subcribing (and if any of it will be missed) and let them decide if they can live with this or not (by throwing panic on their own). If there was no output before late subcriber calls `Watch()` then nothing is lost and things should work as expected.

There are probably other possible solutions.

Reply:
I've designed this system to specifically disallow late subscribers for two reasons:
1. We don't need it now.
2. We don't have an upcoming usecase for it.

Am I missing something here? Is there a use case that you're aware of?

Regarding how `Watch` and `start` are wired into the current steps I've addressed potential fragility in three ways:
1. I've documented the behavior: https://github.com/GoogleCloudPlatform/compute-image-tools/pull/1243/files#diff-364879f77dafcc960afbb3cc4ded0dabR35
2. If someone inadvertently performs a late subscription, they won't receive updates, so it will fail loudly.
3. If someone changes `step_wait_for_instances_signal.go` and moves `Watch` to an incorrect location, then tests will fail (unit and integration)

Regarding your two potential solutions (buffer or indicate that polling has already started), I think those are fantastic options *if* we need to support late subscribers. Do you have a use case in mind, though?

Reply:
In my opinion "Watch" and "start" should be bounded together. No matter whether "start" has been done, it should be called again blindly without side effect.
It's also a reason why we should allow "late subscribe": it may have been started.

It's not a good idea to always rely on create_instance to start the watcher. The reason is that we definitely should be able to wait for a signal from a instance which is not created in the current daisy workflow.